mle/integration/github.py

```diff
@@ -0,0 +1,97 @@
+        :param github_token: the Github token.
+        """
+        self.github_repo = github_repo
+        if not github_token:
+            github_token = os.getenv("GITHUB_TOKEN")
+        self.headers = {
+            "Authorization": f"token {github_token}",
+            "Accept": "application/vnd.github.v3+json"
+        }
+
+    def _make_request(self, endpoint, params=None):
+        """Make a GET request to the GitHub API."""
+        url = f"{self.BASE_URL}/repos/{self.github_repo}/{endpoint}"
+        response = requests.get(url, headers=self.headers, params=params)
+        response.raise_for_status()
+        return response.json()
+
+    def process_source_code(self, file_pattern="*"):
+        """
+        Process source code files in the repository.
+        :param file_pattern: Wildcard pattern to filter files (e.g., "*.py" for Python files)
+        :return: Dictionary with file paths as keys and file contents as values
+        """
+
+        def get_contents(path=""):
+            contents = self._make_request(f"contents/{path}")
+            if isinstance(contents, list):
+                for item in contents:
+                    if item['type'] == 'dir':
+                        yield from get_contents(item['path'])
+                    elif fnmatch(item['name'], file_pattern):
+                        yield item
+            elif fnmatch(contents['name'], file_pattern):
+                yield contents
+
+        source_code = {}
+        for item in get_contents():
+            try:
+                if 'content' in item and item.get('encoding') == 'base64':
+                    content = base64.b64decode(item['content']).decode('utf-8')
+                elif 'download_url' in item:
+                    response = requests.get(item['download_url'], headers=self.headers)
+                    response.raise_for_status()
+                    content = response.text
+                else:
+                    content = f"File too large to process directly. Size: {item.get('size', 'unknown')} bytes"
+                source_code[item['path']] = content
+            except Exception as e:
+                error_message = f"Error: {str(e)}"
+                if isinstance(e, requests.exceptions.RequestException):
+                    error_message += f" (Status code: {e.response.status_code})"
+                print(f"Error processing file {item['path']}: {error_message}")
+                source_code[item['path']] = f"Unable to process content: {error_message}"
+        return source_code
+
+    def process_commit_history(self, limit=10):
+        commits = self._make_request("commits", params={"per_page": limit})
+        commit_history = {}
+        for commit in commits:
+            commit_history[commit['sha']] = {
+                "author": commit['commit']['author']['name'],
+                "date": commit['commit']['author']['date'],
+                "message": commit['commit']['message']
```

Comment:
is it nessary and feasible to get the code diff for a specific commit? maybe the codes have more details (e.g., 'TODO', 'FIXME') for analysis?

Reply:
It makes sense, I can try to add a PR-level diff function into this helper class. We need to have those functions first and then think about how to use them.